### PR TITLE
Avoid deadlocks in Log/TinyLog

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Avoid deadlock when executing INSERT SELECT into itself from a table with `TinyLog` or `Log` table engines. This closes #6802.

### CODE REVIEW

**Comment by akuzm** [comment]
https://github.com/ClickHouse/ClickHouse/issues/14570

Is it going to fix this issue as well?

File: src/Storages/StorageDistributed.cpp [hunk 1]
```diff
@@ -878,7 +878,7 @@ void StorageDistributed::renameOnDisk(const String & new_path_to_table_data)
     {
         const String path(disk->getPath());
         auto new_path = path + new_path_to_table_data;
-        Poco::File(path + relative_data_path).renameTo(new_path);
+        disk->moveDirectory(path + relative_data_path, new_path);
```
Comment (Jokser) [inline src/Storages/StorageDistributed.cpp:881]: You don't need to pass disk->path to disk methods explicitly. Every disk method already adds the disk path to paths given in parameters.

Comment (alexey-milovidov) [reply src/Storages/StorageDistributed.cpp]: Ok.

**Comment by alexey-milovidov** [comment]
test_distributed_over_live_view - known bad test @vzakaznikov

**Comment by alexey-milovidov** [comment]
AST Fuzzer: https://github.com/ClickHouse/ClickHouse/issues/14627

**Review by akuzm — APPROVED** [review]

File: src/Storages/StorageStripeLog.cpp [hunk 2]
```diff
@@ -345,24 +358,28 @@ Pipe StorageStripeLog::read(
...
+    std::shared_lock lock(rwlock, getLockTimeout(context));
+    if (!lock)
+        throw Exception("Lock timeout exceeded", ErrorCodes::TIMEOUT_EXCEEDED);
+
     return file_checker.check();
 }
 
 void StorageStripeLog::truncate(const ASTPtr &, const StorageMetadataPtr &, const Context &, TableExclusiveLockHolder &)
 {
-    std::shared_lock<std::shared_mutex> lock(rwlock);
```
Comment (akuzm) [inline src/Storages/StorageStripeLog.cpp:362]: Why is it not needed anymore?

**Comment by akuzm** [comment]
> #14570
> 
> Is it going to fix this issue as well?

It does fix it.

**Comment by alexey-milovidov** [comment]
@akuzm The new test case is a little bit flawed, could you please make some edits?

**Comment by alexey-milovidov** [comment]
Easy.

**Comment by akuzm** [comment]
Integration test failures reported https://github.com/ClickHouse/ClickHouse/issues/14414
Stress test broken in master.

Yandex check doesn't like distributed test, need to disable it somehow.

File: src/Storages/StorageStripeLog.cpp (hunk 2 above)
Comment (alexey-milovidov) [reply src/Storages/StorageStripeLog.cpp:362]: Truncate, drop, rename are synchronized on higher level.

**Comment by akuzm** [comment]
Yandex check fixed, and other checks passed before, can merge now.

**Comment by alexey-milovidov** [comment]
Too early...
https://clickhouse-test-reports.s3.yandex.net/15240/dce6a436f37713224c242092446e1a9fcf8b4a4a/functional_stateless_tests_(release)/test_run.txt.out.log

**Pull request merged**